cairo/src/lib.rs

```diff
@@ -60,36 +60,77 @@ macro_rules! gvalue_impl {
             }
         }
 
-        impl<'a> glib::value::FromValueOptional<'a> for $name {
-            unsafe fn from_value_optional(v: &'a glib::value::Value) -> Option<Self> {
-                let ptr = glib::gobject_ffi::g_value_get_boxed(v.to_glib_none().0);
-                assert!(!ptr.is_null());
-                from_glib_none(ptr as *mut $ffi_name)
-            }
+        impl glib::value::ValueType for $name {
+            type Type = Self;
         }
 
-        impl glib::value::SetValue for $name {
-            unsafe fn set_value(v: &mut glib::value::Value, s: &Self) {
-                glib::gobject_ffi::g_value_set_boxed(
-                    v.to_glib_none_mut().0,
-                    s.to_glib_none().0 as glib::ffi::gpointer,
-                );
+        impl<'a> glib::value::FromValue<'a> for $name {
+            type Error = glib::value::WrongValueTypeOrNoneError;
+
+            fn check(value: &'a glib::Value) -> Result<(), Self::Error> {
+                glib::value::WrongValueTypeError::check::<$name>(value)?;
+
+                unsafe {
+                    let ptr = glib::gobject_ffi::g_value_get_boxed(
+                        glib::translate::ToGlibPtr::to_glib_none(value).0,
+                    );
+                    if ptr.is_null() {
+                        return Err(glib::value::WrongValueTypeOrNoneError::UnexpectedNone);
+                    }
+                }
+
+                Ok(())
+            }
+
+            fn from_value(value: &'a glib::Value) -> Result<Self, Self::Error> {
+                Self::check(value)?;
```

Comment:
To take this as an example, we could make `from_value()` unsafe again and have it return `Self` and require that `check()` is called before, always.

Opinions?

Reply:
Maybe better. This is not supposed to be called directly anyway and adding the check everywhere introduces quite a bit of boilerplate.